### PR TITLE
sd-dhcp-server: persist hardware addresses in leases

### LINKED ISSUE #33704: DHCP Server does not persist mac addresses associated with leases
### systemd version the issue has been seen with

256.2-1

### Used distribution

Debian 12 (note: I backported systemd 256.2-1 from trixie)

### Linux kernel version used

6.1.0-22-amd64

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

### Expected behaviour you didn't see

MAC Addresses should persist across reboots

### Unexpected behaviour you saw

MAC Addresses get zeroed out when loading them from persistent storage

### Steps to reproduce the problem

1. Configure a DHCP server
2. Have a device get a lease on the network
3. Observe the hwaddr field on the lease is set
4. Reboot (or just restart systemd-networkd)
5. Observe the hwaddr field on the lease is zeroed out

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
# dbus-send --system --dest=org.freedesktop.network1 --print-reply /org/freedesktop/network1/link/4 org.freedesktop.DBus.Properties.Get string:org.freedesktop.network1.DHCPServer string:Leases
method return time=1720723691.260427 sender=:1.99 -> destination=:1.107 serial=37 reply_serial=2
   variant       array [
         struct {
            uint32 2
            array of bytes [
               01 ec b5 fa ad 4d b9
            ]
            array of bytes [
               0a f0 01 77
            ]
            array of bytes [
               00 00 00 00
            ]
            array of bytes [
               ec b5 fa ad 4d b9 00 00 00 00 00 00 00 00 00 00
            ]
            uint64 6182211841
         }
      ]
# # Notice the mac address visible above (field 4) is populated
# systemctl restart systemd-networkd
# dbus-send --system --dest=org.freedesktop.network1 --print-reply /org/freedesktop/network1/link/4 org.freedesktop.DBus.Properties.Get string:org.freedesktop.network1.DHCPServer string:Leases
method return time=1720723717.300620 sender=:1.108 -> destination=:1.109 serial=15 reply_serial=2
   variant       array [
         struct {
            uint32 2
            array of bytes [
               01 ec b5 fa ad 4d b9
            ]
            array of bytes [
               0a f0 01 77
            ]
            array of bytes [
               00 00 00 00
            ]
            array of bytes [
               00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
            ]
            uint64 6182211841
         }
      ]
# # Notice that the mac address (field 4) is now zeroed out :(
# networkctl status brlan
● 4: brlan
                   Link File: /usr/lib/systemd/network/99-default.link
                Network File: /usr/lib/systemd/network/brlan.network
                       State: routable (configured)
                Online state: online
                        Type: bridge
                        Kind: bridge
                      Driver: bridge
            Hardware Address: c2:5a:ce:60:31:b3
                         MTU: 1500 (min: 68, max: 65535)
                       QDisc: noqueue
IPv6 Address Generation Mode: eui64
               Forward Delay: 15s
                  Hello Time: 2s
                     Max Age: 20s
                 Ageing Time: 5min
                    Priority: 32768
                         STP: no
      Multicast IGMP Version: 2
                        Cost: 2000
                  Port State: disabled
    Number of Queues (Tx/Rx): 1/1
            Auto negotiation: no
                       Speed: 100Mbps
                     Address: 10.240.1.1
                              fe80::c05a:ceff:fe60:31b3
           Activation Policy: up
         Required For Online: yes
          DHCPv6 Client DUID: DUID-EN/Vendor:0000ab1193cc7c40f2ad58b9
         Offered DHCP leases: 10.240.1.119 (to ec:b5:fa:ad:4d:b9)

Jul 11 18:47:23 systemd-networkd[2092]: brlan: netdev exists, using existing without changing its parameters
Jul 11 18:47:23 systemd-networkd[2092]: brlan: Configuring with /usr/lib/systemd/network/brlan.network.
Jul 11 18:47:57 systemd-networkd[2092]: brlan: Lost carrier
Jul 11 18:48:07 systemd-networkd[2092]: brlan: Gained carrier
Jul 11 18:48:34 systemd-networkd[2130]: brlan: netdev ready
Jul 11 18:48:34 systemd-networkd[2130]: brlan: Link UP
Jul 11 18:48:34 systemd-networkd[2130]: brlan: Gained carrier
Jul 11 18:48:34 systemd-networkd[2130]: brlan: Gained IPv6LL
Jul 11 18:48:34 systemd-networkd[2130]: brlan: netdev exists, using existing without changing its parameters
Jul 11 18:48:34 systemd-networkd[2130]: brlan: Configuring with /usr/lib/systemd/network/brlan.network.
#
```

### PR DESCRIPTION
**Pull request by Mstrodl**
Fixes #33704

Please let me know if I'm doing something wrong! I'm not thrilled about how I'm storing hlen and chaddr separately, but I figured it would be better to encode it the same way as in the actual struct.

### CODE REVIEW

File: src/libsystemd/sd-json/json-util.c [hunk 1]
```diff
@@ -92,6 +92,22 @@ int json_dispatch_user_group_name(const char *name, sd_json_variant *variant, sd
         return 0;
 }
 
+int json_dispatch_chaddr(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata) {
```
Comment (YHNdnzj) [inline src/libsystemd/sd-json/json-util.c]: Hmm, this is only useful in the libsystemd-network context. Hence move it there?

Comment (Mstrodl) [reply src/libsystemd/sd-json/json-util.c]: I moved this into the lease file instead. Is that what you were thinking?

Comment (Mstrodl) [reply src/libsystemd/sd-json/json-util.c]: On second thought, I made it static too.

**Comment by Mstrodl** [comment]
@mrc0mmand I could be wrong, but it looks like the CI is failing for reasons unrelated to my PR:
```
12:53:28 [cmd_retry] Command 'dnf -y --skip-unavailable builddep systemd' failed (EC: 1) [try 4/5]
12:56:08 enabling baseos-source repository
12:56:08 enabling appstream-source repository
12:56:08 enabling crb-source repository
12:56:08 enabling extras-common-source repository
12:56:08 enabling epel-source repository
12:56:08 enabling epel-cisco-openh264-source repository
12:56:08 enabling epel-next-source repository
12:56:12 CentOS Stream 9 - BaseOS - Source               981  B/s | 3.0 kB     00:03    
12:56:12 Errors during downloading metadata for repository 'baseos-source':
12:56:12   - Downloading successful, but checksum doesn't match. Calculated: e13bd479141ed656a70be1867b0c7258625f469b13baaf70011319723361358415da1f8417f92cf37eaf5dd4bb83dddae4ae7e52fddbc4c6dfb0ac3af2c3233a(sha512)  Expected: c435af47f3197ce3267e5244050b6b1ef1ee668c91cee288d83c887471d5c239b1c82cfec43c283624ba25d40b6eb01e56768a54bd7ffa4b47a5e21fb62365b3(sha512) 
12:56:12 Error: Failed to download metadata for repo 'baseos-source': Cannot download repomd.xml: Cannot download repodata/repomd.xml: All mirrors were tried
12:56:12 [cmd_retry] Command 'dnf -y --skip-unavailable builddep systemd' failed (EC: 1) [try 5/5]
```

**Comment by Mstrodl** [comment]
@YHNdnzj I think all the failing CI is not an issue on my end, but maybe I'm mistaken? Is it usually this inconsistent?
- **CentOS CI (CentOS Stream 9)** Only common failure, but appears to be failing because of problems installing dependencies? [Nobody else seems to be passing this either.](https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-centos9s/)
- **ci/semaphoreci/pr: Debian autopkgtest (LXC)** [Did not fail last time](), appears to be failing for networking reasons, 
- **rpm-build:fedora-rawhide-i386** Fails at the compilation step on code I haven't touched at all? [Did not fail last time](https://github.com/systemd/systemd/pull/33713/checks?check_run_id=27389558193). (Granted, I've rebased since then)
- **CentOS CI (Arch Linux)** [Didn't fail last time](https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-vagrant-archlinux/15841/), spurious failure in some dbus stuff I haven't touched?

**Comment by Mstrodl** [comment]
Just noticed the autopkgtest failures... I can't recall if those succeeded before or not, but I'll take a look at them tomorrow.

**Comment by Mstrodl** [comment]
autopkgtest failures seem to be similarly flaky... amd64 and arm64 have distinct sets of failed tests (in other words, all the tests failed on amd64 passed on arm64 and vice-versa). pp64el appears to have failed for infrastructure-related reasons (nova quota exceeded?) and I'm unsure about s390x, I forgot to look before I pushed.

In any case, it seems like the test suite is flaky or I have some big underlying mistake in my code, which is possible, but seems a little unlikely given the nature of the failures.

I suspect if I could retry individual tests they would all pass.

**Comment by Mstrodl** [comment]
@bluca Hey, can you take a look at this when you get a chance? (Or somebody else? idk)

**Review by bluca — APPROVED** [review]

**Comment by yuwata** [comment]
Belated LGTM. Thank you.

**Comment by Mstrodl** [comment]
> Belated LGTM. Thank you.

:shipit:

**Pull request merged**